app/src/main/java/com/duckduckgo/app/bookmarks/model/SavedSitesRepository.kt

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.bookmarks.model
+
+import com.duckduckgo.app.bookmarks.model.SavedSite.Bookmark
+import com.duckduckgo.app.bookmarks.model.SavedSite.Favorite
+import com.duckduckgo.app.global.DefaultDispatcherProvider
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.sync.store.Entity
+import com.duckduckgo.sync.store.EntityType.BOOKMARK
+import com.duckduckgo.sync.store.EntityType.FOLDER
+import com.duckduckgo.sync.store.Relation
+import com.duckduckgo.sync.store.SyncEntitiesDao
+import com.duckduckgo.sync.store.SyncRelationsDao
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+
+interface SavedSitesRepository {
+
+    // getFavorites and getFolderContent will be the same once we can merge Favorites / Bookmarks
+
+    suspend fun getFolderContent(parentId: Long?): Flow<Pair<List<Bookmark>, List<BookmarkFolder>>>
+
+    fun getBookmarks(): Flow<List<Bookmark>>
+
+    fun getBookmark(url: String): Bookmark?
+
+    fun getFavorites(): Flow<List<Favorite>>
+
+    fun getFavorite(url: String): Favorite?
+
+    fun hasBookmarks(): Boolean
+
+    fun insertBookmark(
+        url: String,
+        title: String,
+    ): Bookmark
+
+    fun insertFavorite(
+        url: String,
+        title: String,
+    ): Favorite
+
+    fun insert(savedSite: SavedSite)
+    fun delete(savedSite: SavedSite)
+    fun update(savedSite: SavedSite)
+
+    fun updateWithPosition(favorites: List<Favorite>)
+
+    fun insert(folder: BookmarkFolder)
+    fun update(folder: BookmarkFolder)
+
+    fun delete(folder: BookmarkFolder)
+
+    fun getFolder(folderId: Long): BookmarkFolder
+}
+
+class RealSavedSitesRepository(
```

Comment:
Thinking out loud...should we protect no one creates a folder with `favorites_root` or `bookmarks_root` as names?

Reply:
the conflict resolution is `OnConflictStrategy.REPLACE` and the primaryKey is 'entityId'. if someone adds an Entity folder with that id, is not a user, so I'd guess it was intentional.